src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,174 @@
+            add(value);
+            return;
+        }
+
+        Node<T> current = getNodeByIndex(index);
+        newNode.next = current;
+        newNode.prev = current.prev;
+        current.prev = newNode;
+
+        if (current == head) {
+            head = newNode;
+        } else {
+            newNode.prev.next = newNode;
+        }
+
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNodeByIndex(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+
+        Node<T> current = getNodeByIndex(index);
+        T value;
+
+        if (size == 1) {
+            value = head.value;
+            clear();
+            return value;
+        }
+
+        value = current.value;
+
+        if (current == tail) {
+            current.prev.next = null;
+            tail = current.prev;
+        } else if (current == head) {
+            current.next.prev = null;
+            head = current.next;
+        } else {
+            current.prev.next = current.next;
+            current.next.prev = current.prev;
+        }
+
+        size--;
+        return value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        for (int i = 0; i < size; i++) {
+            if (current.value == null
+                    && object == null
+                    || object != null
+                    && object.equals(current.value)) {
+                remove(i);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void clear() {
+        head = tail = null;
+        size = 0;
+    }
+
+    private Node<T> getNodeFromHead(int index) {
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current;
+    }
+
+    private Node<T> getNodeFromTail(int index) {
+        Node<T> current = tail;
+        for (int i = size - 1; i > index; i--) {
+            current = current.prev;
+        }
+        return current;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        if (index < size / 2) {
+            return getNodeFromHead(index);
+        } else {
+            return getNodeFromTail(index);
+        }
+    }
```

Comment:
Instead of these three methods, I use one method, but I am not sure that my decision is correct
`private Node<T> getNodeByIndex(int index) {
        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException("Unable to set index: " + index);
        }
        Node<T> current = head;
        for (int i = 0; i < index; i++) {
            current = current.next;
        }
        return current;
    }`